File: W3/homeWork_W3/С++/Gun/Gun.cpp
#include "Gun.h"

#define BULLET 1 
#define SHOOT 1

using namespace std;

Gun::Gun(const string& model, int capacity) {
    this->amount = 0;
    this->capacity = capacity;
    this->isReady = false;
    this->model = model;
    this->totalShots = 0;
}

Gun::~Gun() {
    
}

int Gun::getAmount() const {
    return this->amount;    
}

int Gun::getCapacity() const {
    return this->capacity;
}

bool Gun::ready() const {
    return this->isReady;
}

const string& Gun::getModel() const {
    return this->model;
}

int Gun::getTotalShots() const {
    return this->totalShots;
}

void Gun::prepare() {
    this->isReady = !this->isReady;
}

void Gun::reload() {
    if ( this->capacity <= 0 ) {
        cout << "You entered incorrect magazine capacity!" << endl;
        return;
    }
    this->amount = this->capacity;
}

void Gun::shoot() {
    if ( !this->ready() ) {
        throw NotReady();
    }
    if ( this->amount == 0 ) {
        throw OutOfRounds();
    }
    cout << "Bang!" << endl;
    this->amount -= BULLET;
    this->totalShots += SHOOT;
}

ostream& operator<<(ostream& out, const Gun& gun) {
    out << "Characteristics of the gun:\n" 
        << "The gun model - " << gun.getModel() << ".\n"
        << "Ready to shoot - " << gun.ready() << ".\n"
        << "Bullets quantity - " << gun.getAmount() << ".\n"
        << "Magazine capacity - " << gun.getCapacity() << ".\n"
        << "Total shots - " << gun.getTotalShots() << ".";
        
        return out;   
}

File: W3/homeWork_W3/Java/Gun/Gun.java
public class Gun {
    private int amount;
    private int capacity;
    private boolean isReady;
    private String model;
    private int totalShots;
    private final int BULLET = 1;
    private final int SHOOT = 1;
    
    public Gun(String model, int capacity) {
        this.amount = 0;
        this.capacity = capacity;
        this.isReady = false;
        this.model = model;
        this.totalShots = 0;
    }
        
    public Gun() {
        this("Beretta", 8);
    }
    
    public final int getAmount() {
        return this.amount;    
    }

    public final int getCapacity() {
        return this.capacity;
    }

    public final boolean ready() {
        return this.isReady;
    }

    public final String getModel() {
        return this.model;
    }

    public final int getTotalShots() {
        return this.totalShots;
    }

    public void prepare() {
        this.isReady = !this.isReady;
    }

    public void reload() {
        if ( this.capacity <= 0 ) {
            System.out.println("You entered incorrect magazine capacity!");
            return;
        }
        this.amount = this.capacity;
    }
    
    public void shoot() throws NotReadyException, OutOfRoundsException {
        if ( !this.ready() ) {
            throw new NotReadyException();
        }
        if ( this.amount == 0 ) {
            throw new OutOfRoundsException();
        }
        System.out.println("Bang!");
        this.amount -= BULLET;
        this.totalShots += SHOOT;
    }
    
    @Override
    public String toString() {
        return "Characteristics of the gun:\n" 
            + "The gun model - " + this.model + ".\n"
            + "Ready to shoot - " + this.isReady + ".\n"
            + "Bullets quantity - " + this.amount + ".\n"
            + "Magazine capacity - " + this.capacity + ".\n"
            + "Total shots - " + this.totalShots + ".";
    }
}
